### PR TITLE
Week 1

### PR DESCRIPTION
**Pull request by jitenvij1992**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -42,6 +42,21 @@
 			<artifactId>h2</artifactId>
 			<scope>runtime</scope>
 		</dependency>
+		<!-- https://mvnrepository.com/artifact/com.google.guava/guava -->
+		<dependency>
+			<groupId>com.google.guava</groupId>
+			<artifactId>guava</artifactId>
+			<version>27.0.1-jre</version>
```
Comment (vinaylakkam) [inline pom.xml]: Did you use this? Most of the guava features are available with Java8, so you can safely delete I believe

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        this.productDetailService = productDetailService;
+        this.productDeleteService = productDeleteService;
+    }
+
+    @GetMapping(value = "/", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<Product>> getAllProducts() {
+        return ResponseEntity.ok(productDetailService.getAvailableProducts());
+    }
+
+    @PostMapping(value = "/create", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: `/create` is not a suggested convention. `POST /products` will do. Please check this: [http://microformats.org/wiki/rest/urls ](http://microformats.org/wiki/rest/urls )

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    public ResponseEntity<String> insertProduct(@RequestBody Product payload) {
+
+        logger.info("Payload received to insert data in product service with value {}", payload);
+        productInsertService.processRawData(payload);
+        return ResponseEntity.ok("Successfully added product in repository");
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: Please check the appropriate status codes for CRUD http requests. In the above case `201(Created)` suits more.

File: src/main/java/com/jpop/productservice/dao/ProductDeleteRepository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.jpop.productservice.dao;
+
+import com.jpop.productservice.model.Product;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ProductDeleteRepository extends CrudRepository<Product, Long> {
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/dao/ProductDeleteRepository.java:8]: `CrudRepository` will do here, but please be aware that `JPARepository` has more extension methods. Please explore the differences for learning.

File: src/main/java/com/jpop/productservice/exception/ProductDeletionException.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.jpop.productservice.exception;
+
+public class ProductDeletionException extends RuntimeException {
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/exception/ProductDeletionException.java]: Not mandatory; But can this be `ProductNotFoundException` , so we can use in update and delete scenarios

File: src/main/java/com/jpop/productservice/service/impl/ProductDeleteServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    private ProductDeleteRepository productDeleteRepository;
+
+    @Override
+    public void deleteProduct(long id) {
+        logger.info("Servicing the request to delete product having id {}", id);
+        if(productDeleteRepository.findById(id).isPresent()) {
+            productDeleteRepository.deleteById(id);
+        } else {
+            throw new ProductDeletionException("Product id doesn't exist.");
+        }
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/service/impl/ProductDeleteServiceImpl.java]: This is correct. Also, Optional has some methods to write the code in more declarative way. Check: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html. 

```
productDeleteRepository.findById(id)
   .map(product -> productDeleteRepository.deleteById(id))
   .orElseThrow( ()->new ProductDeletionException(".."))
```

File: src/main/java/com/jpop/productservice/service/impl/ProductDetailServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    public Product getProductDetails(long id) {
+
+        logger.info("Inside Product detail. Getting details for the product having id {}", id);
+        Optional<Product> dataById = productDetailRepository.findById(id);
+
+        if(dataById.isPresent()) {
+            return dataById.get();
+        } else {
+            throw new InvalidProductDetailException("No product found having id "+id);
+        }
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/service/impl/ProductDetailServiceImpl.java]: Please try to use Optional methods like..
```
return productDetailRepository.findById(id)
   .orElseThrow(new InvalidProductDetailException("...")
```

[https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html)

File: src/main/java/com/jpop/productservice/service/impl/ProductInsertServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public void processUpdatedData(Product product, long id) {
+        if(!productInsertRepository.findById(id).isPresent())
+            throw new ProductValidationException("No Product found having id: " +id);
+        product.setId(id);
+        processRawData(product);
+    }
+
+    private boolean validateRawData(Product product) {
+        return Stream.of(product.getName(), product.getPrice(), product.getDescription()).anyMatch(Objects::isNull);
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/service/impl/ProductInsertServiceImpl.java]: Nice.

File: src/test/java/com/jpop/productservice/controller/ProductServiceControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+               .andExpect(status().isOk());
+    }
+
+    @Test
+    public void deleteProductByIdWhenException() throws Exception {
+        Mockito.doThrow(ProductDeletionException.class).when(productDeleteService).deleteProduct(8);
+        mockMvc.perform(MockMvcRequestBuilders.delete("/api/v1/products/8")
+                .contentType(MediaType.APPLICATION_JSON)).
+                andExpect(status().isNoContent());
+    }
```
Comment (vinaylakkam) [inline src/test/java/com/jpop/productservice/controller/ProductServiceControllerTest.java]: Nice unit tests for controller. Did you try testing `update`? We can also write integration tests for spring boot apps. Check this https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-testing.html and try in week2 or later

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.List;
+
+@RestController
+@RequestMapping(value = "/api/v1/products")
+public class ProductServiceController {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProductServiceController.class);
+    private ProductInsertService productInsertService;
+    private ProductDetailService productDetailService;
+    private ProductDeleteService productDeleteService;
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: I felt it's an overhead to write different classes for each CRUD operation. Is this style recommended somewhere?

**Review by vinaylakkam — COMMENTED** [review]
Jiten, It's pretty much comprehensive. I see you followed clean code principles too. Good job. I added some comments. Please let me know if you need any clarification.

File: pom.xml [hunk 11]
```diff
@@ -55,6 +61,14 @@
...
 				<artifactId>spring-boot-maven-plugin</artifactId>
 			</plugin>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-compiler-plugin</artifactId>
+				<configuration>
+					<source>9</source>
+					<target>9</target>
+				</configuration>
+			</plugin>
```
Comment (vinaylakkam) [inline pom.xml]: This plugin is already defined in spring boot parent pom. I think you just need to change `<java.version>1.8</java.version>`  in line no.18 to change the java version

File: pom.xml (hunk 1 above)
Comment (vinaylakkam) [reply pom.xml]: guava is removed now

File: pom.xml [hunk 12]
```diff
@@ -42,6 +42,12 @@
 			<artifactId>h2</artifactId>
 			<scope>runtime</scope>
 		</dependency>
+		<!-- https://mvnrepository.com/artifact/com.google.code.gson/gson -->
+		<dependency>
+			<groupId>com.google.code.gson</groupId>
+			<artifactId>gson</artifactId>
+			<version>2.8.5</version>
+		</dependency>
```
Comment (vinaylakkam) [inline pom.xml:50]: spring internally uses Jackson which has greater annotation support. Do you see any advantages of using gson over the spring boot's default jackson library? I am just asking. Please reply.

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java (hunk 10 above)
Comment (vinaylakkam) [reply src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: Can you please answer this.

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        productInsertService.processRawData(payload);
+        return ResponseEntity.status(HttpStatus.CREATED).body("Successfully added product in repository");
+    }
+
+    @GetMapping(value = "/{id}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> getProductDetails(@PathVariable long id) {
+
+        logger.info("Received request to get details of product having id {}", id);
+        String productData = new Gson().toJson(productDetailService.getProductDetails(id));
+        return ResponseEntity.ok(productData);
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/controller/ProductServiceController.java:63]: Not important. Better to send 204 status in these cases; Not important though

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    public ResponseEntity<List<Product>> getAllProducts() {
+        return ResponseEntity.ok(productDetailService.getAvailableProducts());
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> insertProduct(@RequestBody Product payload) {
+
+        logger.info("Payload received to insert data in product service with value {}", payload);
+        productInsertService.processRawData(payload);
+        return ResponseEntity.status(HttpStatus.CREATED).body("Successfully added product in repository");
```
Comment (vinaylakkam) [inline src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: It is suggested to send the URI (location) in the header back to the user to access the created resource. Try to use `ResponseEntity.created(uri of the resource).body(product object which is created)`. The intention is that the client of this endpoint has to know the id of the product inserted.

File: src/test/java/com/jpop/productservice/ProductServiceApplicationTests.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+//
+//@RunWith(SpringRunner.class)
+//@SpringBootTest
+//public class ProductServiceApplicationTests {
+//
+//	@Test
+//	public void contextLoads() {
+//	}
+//
+//}
```
Comment (vinaylakkam) [inline src/test/java/com/jpop/productservice/ProductServiceApplicationTests.java]: Please delete this if unused.

File: src/test/java/com/jpop/productservice/integration/ProductServiceApplicationIT.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        HttpEntity httpEntity = new HttpEntity(null, headers);
+
+        ResponseEntity<String> responseEntity = testRestTemplate.exchange(createUrl(), HttpMethod.GET, httpEntity, String.class);
+        assertEquals("Status code is invalid", 200, responseEntity.getStatusCode().value());
+    }
+
+    private String createUrl() {
+        return "http://localhost:" + port + "/api/v1/products/";
+    }
+}
```
Comment (vinaylakkam) [inline src/test/java/com/jpop/productservice/integration/ProductServiceApplicationIT.java:40]: Good that you have written an integration test for `GET products`. Please add a task in week2 or week3 to complete this for other methods

**Review by vinaylakkam — CHANGES_REQUESTED** [review]
Please check the minor comments. Please reply to the comments, if you have resolved something or have something to say.

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java (hunk 10 above)
Comment (jitenvij1992) [reply src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: I have read one article and according to that class should not have more than 150 lines and crud operations can be done in separate class.

File: pom.xml (hunk 12 above)
Comment (jitenvij1992) [reply pom.xml:50]: There are so many methods in gson which are helpful. Earlier I used some methods to convert the complex list data and also retrieving some data from collections

File: src/main/java/com/jpop/productservice/controller/ProductServiceController.java (hunk 14 above)
Comment (jitenvij1992) [reply src/main/java/com/jpop/productservice/controller/ProductServiceController.java]: Thank you so much for the suggestion on this.

File: src/test/java/com/jpop/productservice/integration/ProductServiceApplicationIT.java (hunk 16 above)
Comment (jitenvij1992) [reply src/test/java/com/jpop/productservice/integration/ProductServiceApplicationIT.java:40]: Sure

File: src/test/java/com/jpop/productservice/ProductServiceApplicationTests.java (hunk 15 above)
Comment (jitenvij1992) [reply src/test/java/com/jpop/productservice/ProductServiceApplicationTests.java]: done

File: pom.xml (hunk 11 above)
Comment (jitenvij1992) [reply pom.xml]: yeah, I changed the version now

File: src/test/java/com/jpop/productservice/controller/ProductServiceControllerTest.java (hunk 9 above)
Comment (jitenvij1992) [reply src/test/java/com/jpop/productservice/controller/ProductServiceControllerTest.java]: yeah, I tried and wrote

File: pom.xml (hunk 12 above)
Comment (vinaylakkam) [reply pom.xml:50]: Ok. I have read through some article. jackson+spring evolved a lot. So, for week3 assignment use jackson. so you will have better understanding of both.

**Review by vinaylakkam — APPROVED** [review]
Nice job. Please merge.

**Pull request merged**